File: tools/3Dgame-design/src/stair-profile.mjs
// Profile coordinates are Blender Z-up, with the origin at the upper endpoint.
export function stairProfileParts(profile, settings) {
  const count = Math.max(4, Math.round(settings.stepCount));
  const run = Math.max(3, Number(settings.runLength));
  const rise = run * Math.tan(Math.max(5, Math.min(60, settings.slopeDeg)) * Math.PI / 180);
  const parts = [];
  for (const part of profile.parts) {
    if ((settings.omitParts || []).includes(part.name)) continue;
    const repeats = part.role === 'tread' ? count : 1;
    for (let i = 0; i < repeats; i++) {
      const vertices = part.vertices.map(([x, y, z]) => {
        if (part.role === 'tread') {
          return [x * (settings.width - .42) / 5.58,
            (i + .5) * run / count + .125 + (y - .5) * (run / count - .03) / .72,
            z - i * rise / count];
        }
        return [x * settings.width / 6, y * run / 15,
          z + y * .4 - y / 15 * rise + (part.role === 'guard' ? settings.handrailHeight - 1.2 : 0)];
      });
      parts.push({ name: part.role === 'tread' ? `踏步_${i + 1}` : part.name, vertices, faces: part.faces });
    }
  }
  return parts;
}

File: tools/3Dgame-design/blender/stair_profile.py
import json
import math
from pathlib import Path

def profile_parts(settings):
    profile = json.loads((Path(__file__).resolve().parent.parent / 'src/stair-flight-profile.json').read_text())
    count = max(4, round(settings['stepCount']))
    run = max(3, settings['runLength'])
    rise = run * math.tan(math.radians(max(5, min(60, settings['slopeDeg']))))
    result = []
    for part in profile['parts']:
        if part['name'] in settings.get('omitParts',[]): continue
        for i in range(count if part['role'] == 'tread' else 1):
            vertices = []
            for x,y,z in part['vertices']:
                if part['role'] == 'tread':
                    vertices.append((x*(settings['width']-.42)/5.58,(i+.5)*run/count+.125+(y-.5)*(run/count-.03)/.72,z-i*rise/count))
                else:
                    vertices.append((x*settings['width']/6,y*run/15,z+y*.4-y/15*rise+(settings['handrailHeight']-1.2 if part['role']=='guard' else 0)))
            result.append(dict(name=f'踏步_{i+1}' if part['role']=='tread' else part['name'],vertices=vertices,faces=part['faces']))
    return result
